### ISSUE TITLE
Add "Difficulty" to Settings

### ISSUE THREAD

**Issue by sanicki** [sanicki/chord-quiz#48]
<html><head></head><body><div _ngcontent-ng-c1840272799="" class="markdown markdown-main-panel tutor-markdown-rendering stronger enable-updated-hr-color" style="--animation-duration: 400ms; --fade-animation-function: linear; font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;" id="model-response-message-contentr_16d5b3fafa030f22" aria-live="polite" aria-busy="false" dir="ltr"><p data-path-to-node="0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Building an app that balances accuracy with beginner-friendly lenience is a classic signal processing challenge. Since acoustic string instruments can have buzzing strings, muted notes, or slight tuning issues, you need to shift from "exact matching" to "probabilistic matching."</p><p data-path-to-node="1" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Here is how you can implement a robust detection system that feels encouraging rather than punishing.</p><hr data-path-to-node="2" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><h2 data-path-to-node="3" style="font-family: Google Sans, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">1. Use Pitch Class Profiling (Chromagram)</h2><p data-path-to-node="4" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Instead of looking for raw frequencies, convert the audio signal into a <b data-path-to-node="4" data-index-in-node="72" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Chroma feature</b> (or Chromagram). A Chromagram maps the entire spectrum of sound into 12 bins representing the 12 semitones of the musical scale.</p><ul data-path-to-node="5" style="padding-inline-start: 32px; font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><li style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="5,0,0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="5,0,0" data-index-in-node="0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Why it works:</b> It ignores the octave and focuses on the "color" of the chord. If a beginner hits a G chord but their finger mutes the high E string, the G, B, and D bins will still show high energy, allowing you to identify the G major "signature."</p></li><li style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="5,1,0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="5,1,0" data-index-in-node="0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">The Tolerance:</b> You can set a threshold where a chord is "correct" if the primary notes (the Root, Third, and Fifth) exceed a certain energy level relative to the background noise.</p></li></ul><hr data-path-to-node="6" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><h2 data-path-to-node="7" style="font-family: Google Sans, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">2. Implement Fuzzy Logic for "Finger Buzz"</h2><p data-path-to-node="8" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Beginners often don't press hard enough, leading to "buzzing" or inharmonic noise.</p><ul data-path-to-node="9" style="padding-inline-start: 32px; font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><li style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="9,0,0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="9,0,0" data-index-in-node="0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Harmonic Product Spectrum (HPS):</b> This algorithm helps identify the fundamental frequency (<span class="math-inline" data-math="f_0" data-index-in-node="90" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">$f_0$</span>) even when the signal is messy.</p></li><li style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="9,1,0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="9,1,0" data-index-in-node="0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Confidence Scores:</b> Rather than a binary "True/False," have your engine return a confidence value (0.0 to 1.0).</p><ul data-path-to-node="9,1,1" style="padding-inline-start: 32px; font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><li style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="9,1,1,0,0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="9,1,1,0,0" data-index-in-node="0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">90%+:</b> Perfect.</p></li><li style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="9,1,1,1,0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="9,1,1,1,0" data-index-in-node="0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">70%–89%:</b> "Close enough"—trigger a success animation.</p></li><li style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="9,1,1,2,0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="9,1,1,2,0" data-index-in-node="0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Below 60%:</b> Likely a different chord or too much deadening.</p></li></ul></li></ul><hr data-path-to-node="10" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><h2 data-path-to-node="11" style="font-family: Google Sans, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">3. Temporal Smoothing (The "Persistence" Rule)</h2><p data-path-to-node="12" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Beginners take a moment to settle into a chord. If you check for a perfect match at a single millisecond, you’ll likely fail them during the "strum attack" phase.</p><ul data-path-to-node="13" style="padding-inline-start: 32px; font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><li style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="13,0,0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="13,0,0" data-index-in-node="0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">The Windowing Method:</b> Require the user to hold the chord for a specific duration (e.g., 200–400ms) before confirming it.</p></li><li style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="13,1,0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="13,1,0" data-index-in-node="0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Averaging:</b> Take the average of the last 5–10 audio frames. If the average signal matches a C Major, confirm it even if one frame was a "miss."</p></li></ul><hr data-path-to-node="14" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><h2 data-path-to-node="15" style="font-family: Google Sans, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">4. Prioritize the "Flavor" Notes</h2><p data-path-to-node="16" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">In music theory, some notes are more important than others for identification.</p><ul data-path-to-node="17" style="padding-inline-start: 32px; font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><li style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="17,0,0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="17,0,0" data-index-in-node="0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Root and Third:</b> To distinguish between C Major and C Minor, the "Third" (the E or Eb) is vital.</p></li><li style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="17,1,0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="17,1,0" data-index-in-node="0" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">The "Ignore" List:</b> You can program the app to be lenient on the 5th (the G in a C chord) or the doubling of octaves, as these are often the notes beginners accidentally mute.</p></li></ul><hr data-path-to-node="18" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><h2 data-path-to-node="19" style="font-family: Google Sans, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">5. Technical Implementation Path</h2><p data-path-to-node="20" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">For an Android environment, you don't want to build this from scratch. Use established libraries to handle the heavy lifting:</p>
Component | Recommendation
-- | --
Audio Processing | TarsosDSP (Java/Kotlin) or Aubio (C/C++ via NDK). They have built-in pitch detection and FFT (Fast Fourier Transform) classes.
Model | TensorFlow Lite. You can train a simple neural network on "clean" vs "messy" beginner chords to teach the app what a "lazy C major" sounds like.
Latency | Use the Oboe library (C++) to ensure low-latency audio monitoring on Android, which is crucial for real-time feedback.

<h3 data-path-to-node="22" style="font-family: Google Sans, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">A Pro Tip for UX:</h3><p data-path-to-node="23" style="font-family: Google Sans Text, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Instead of saying "Wrong Chord," use visual cues. If they are playing a G Major but missing the B string, show a UI element that gently pulses on the string they are missing. This turns your app from a "judge" into a "tutor."</p></div></body></html>
